### PR TITLE
Initial addition of CustomBuild and local STI testing

### PR DESCRIPTION
**Pull request by mfojtik**

### CODE REVIEW

**Comment by soltysh** [comment]
LGTM

File: image_writers_guide/sti_testing.adoc [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+== Basic work flow
+
+This section describes the default work flow you should be using for testing the
+STI image builders:
+
+1. docker build -t BUILDER_IMAGE .
+1. docker run BUILDER_IMAGE <verify the usage script is working>
+1. sti build file:///path-to-sample-app BUILDER_IMAGE APPLICATION_IMAGE
+1. <in case you support save-artifacts, execute 2. step again to verify that
+   restore-artifacts works properly>
```
Comment (soltysh) [inline image_writers_guide/sti_testing.adoc]: There's no `restore-artifacts` actually, this is `assemble` responsibility to do that. So I'd propose to write it differently to explain that.

Comment (mfojtik) [inline image_writers_guide/sti_testing.adoc]: good point :-)

Comment (CowboysFan) [inline image_writers_guide/sti_testing.adoc]: @mfojtik Has the change been implemented? Is this ready to be merged?

Comment (mfojtik) [inline image_writers_guide/sti_testing.adoc]: This is ready to merge, the change is already there. See this Trello card: https://trello.com/c/hCMretlK/285-2-document-how-to-setup-sti-build-framework-imageeco-beta2

File: image_writers_guide/sti_testing.adoc [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+This document covers how as a STI builder image author you can test your
+STI image locally and how you can use the OpenShift build system for automated
+testing and continuous integration.
+
+Read more about the link:sti[STI architecture and requirements] before reading
+any further.
+
+== The Goal
+
+As described in the STI documentation, STI does require 'assemble' and 'run'
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: s/does require/requires/

File: image_writers_guide/sti_testing.adoc [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+commands to be present in order to successfully execute the STI build.  You can
+also provide the 'save-artifacts' command for archiving the build artifacts and
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: archiving/reusing

File: image_writers_guide/sti_testing.adoc [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+'usage' command that prints the usage to console when someone runs the Docker
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: the 'usage' command

File: image_writers_guide/sti_testing.adoc [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+image outside the STI.
+
+The goal of testing the STI image is to make sure that all commands described
+above works properly, even if the base Docker image changed or the tooling that
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: work

File: image_writers_guide/sti_testing.adoc [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+invoked by the OpenShift STI image builder and it could be a simple Bash script
+or a static Go binary.
+
+Since the 'test/run' command will perform the STI build, you should have STI
+binary available in your system. See the STI
+https://github.com/openshift/source-to-image/blob/master/README.md#installation[README]
+file and follow the installation instructions.
+
+Since the STI is putting together the application source code and Docker image,
+in order to test it, we would need a sample application source we can use during
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: s/would//

File: image_writers_guide/sti_testing.adoc [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+the test to verify the application source gets successfully converted into a
+Docker image. The sample application should be simple, but it should also
+exercise the 'assemble' command.
+
+== Using sti create
+
+The STI tooling comes with powerful generation tools to speed up the process of
+a new STI image creation. The
+https://github.com/openshift/source-to-image/blob/master/docs/cli.md#sti-create:[sti
+create] will produce all necessary STI scripts and testing tools along with the
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: s/will/command will/

File: image_writers_guide/sti_testing.adoc [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+One thing you need to keep in mind is that the 'test/run' script produced by
+'sti create' command require the sample application sources inside the
+'test/test-app' directory.
+
+== Testing locally
+
+The easiest way to run the STI image tests locally is to use the generated
+Makefile. In case you did not use the 'sti create' command, you can grab this
+template and replace the IMAGE_NAME with your image name.
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: grab what template?  you mean the example that follows this paragraph?

File: image_writers_guide/sti_testing.adoc [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+---
+
+== Basic work flow
+
+This section describes the default work flow you should be using for testing the
+STI image builders:
+
+1. docker build -t BUILDER_IMAGE .
+1. docker run BUILDER_IMAGE <verify the usage script is working>
+1. sti build file:///path-to-sample-app BUILDER_IMAGE APPLICATION_IMAGE
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: OUTPUT_APPLICATION_IMAGE

File: image_writers_guide/sti_testing.adoc [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+.PHONY: test
+test:
+	docker build -t $(IMAGE_NAME)-candidate .
+	IMAGE_NAME=$(IMAGE_NAME)-candidate test/run
+---
+
+== Basic work flow
+
+This section describes the default work flow you should be using for testing the
+STI image builders:
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: is it worth delineating which of these steps are done by the test script, and which are not?

for example docker build -t BUILDER_IMAGE.  is not run by the test script, but the rest is, right?

File: image_writers_guide/sti_testing.adoc [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+1. docker run APPLICATION_IMAGE
+1. <verify the container is running and the application is responding>
+
+Executing all the steps above, should be usually enough to tell if the STI
+commands operates properly.
+
+== Using OpenShift Build for Automated Testing
+
+Another way how you can execute the STI image tests is to use OpenShift platform
+itself as a continuous integration system. OpenShift know how to build the
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: s/know/knows/

File: image_writers_guide/sti_testing.adoc [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+   restoring artifacts works properly>
+1. docker run APPLICATION_IMAGE
+1. <verify the container is running and the application is responding>
+
+Executing all the steps above, should be usually enough to tell if the STI
+commands operates properly.
+
+== Using OpenShift Build for Automated Testing
+
+Another way how you can execute the STI image tests is to use OpenShift platform
```
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: s/how//

**Comment by bparees** [comment]
otherwise lgtm.

File: image_writers_guide/sti_testing.adoc (hunk 8 above)
Comment (mfojtik) [inline image_writers_guide/sti_testing.adoc]: yes

File: image_writers_guide/sti_testing.adoc (hunk 10 above)
Comment (mfojtik) [inline image_writers_guide/sti_testing.adoc]: I should opt-out this step, yeah. The test script assume that there is an image to test built already.

File: image_writers_guide/sti_testing.adoc (hunk 8 above)
Comment (bparees) [inline image_writers_guide/sti_testing.adoc]: then I think  saying "copy the template below" would be clearer

**Comment by mfojtik** [comment]
@bparees thanks! fixed all typos and comments.

**Comment by adellape** [comment]
Thanks for this, @mfojtik. I'll go ahead and merge now, then add it to the build_cfg / make any style edits in a follow-up PR.

**Pull request merged**